### PR TITLE
Don't diff memoized host components in completion phase

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/issues/13425.

View without whitespace: https://github.com/facebook/react/pull/13423/files?w=1

The bigger issue (https://github.com/facebook/react/issues/13424) is still a problem but it has always been broken (I checked as far as 0.11). Here I'm only fixing the regression in React 16 that made this a problem even when `setState` calls are in different components.

In particular, we shouldn't be diffing host components that are outside of `setState` path.

Some indentation changes are due to tests I tweaked but they're unrelated. The only change there is to consistently add and remove `container` from the document. The real changes are only in the newly added test at the bottom, and in `CompleteWork`.

### CODE REVIEW

File: packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap [hunk 1]
```diff
@@ -357,8 +357,8 @@ exports[`ReactDebugFiberPerf skips parents during setState 1`] = `
 
 ⚛ (Committing Changes)
   ⚛ (Committing Snapshot Effects: 0 Total)
-  ⚛ (Committing Host Effects: 6 Total)
-  ⚛ (Calling Lifecycle Methods: 6 Total)
```
Comment (gaearon) [inline packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap]: Should have spotted this earlier. Sigh.

Comment (nhunzaker) [reply packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap]: Is this just way more efficient now?

**Comment by react-linter** [comment]
<details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: d14e443d6e97f6f4ca80d2e9a80300690b71e3c4...4247738695bf02419c53c02629308162eaad6973</p>


  
## react-noop-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-noop-renderer.development.js | +2.1% | +1.5% | 22.88 KB | 23.36 KB | 5.25 KB | 5.33 KB | NODE_DEV
**react-noop-renderer.production.min.js** | **:small_red_triangle:+1.6%** | **:small_red_triangle:+1.9%** | **8.62 KB** | **8.75 KB** | **2.98 KB** | **3.03 KB** | **NODE_PROD**
react-noop-renderer-persistent.development.js | +2.1% | +1.5% | 23.01 KB | 23.48 KB | 5.26 KB | 5.34 KB | NODE_DEV
**react-noop-renderer-persistent.production.min.js** | **:small_red_triangle:+1.5%** | **:small_red_triangle:+1.8%** | **8.64 KB** | **8.77 KB** | **2.98 KB** | **3.04 KB** | **NODE_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-reconciler/src/ReactFiberCompleteWork.js [hunk 2]
```diff
@@ -361,6 +361,10 @@ function completeWork(
         // If we have an alternate, that means this is an update and we need to
         // schedule a side-effect to do the updates.
         const oldProps = current.memoizedProps;
+        if (oldProps === newProps) {
+          break;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberCompleteWork.js]: Be careful with those early returns. I think you’re skipping the ref changing.

File: packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap (hunk 1 above)
Comment (gaearon) [reply packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap]: I think not — I think it was still overcounting actual updates (not entirely sure why). But it's better.

File: packages/react-reconciler/src/ReactFiberCompleteWork.js (hunk 2 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberCompleteWork.js]: Can it even change if we bail out?

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberCompleteWork.js]: I'll push up a fix tho.

Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberCompleteWork.js]: In theory we can have a different element with the same props object.

**Pull request merged**